File: cpp/n387_FirstUniq.cpp
// https://leetcode-cn.com/problems/first-unique-character-in-a-string/
#include <iostream>
#include <string>
#include <map>
#include <vector>
using namespace std;

class Solution {
public:
    int firstUniqChar(string s) {
        map<char, int> count;
        for (size_t i = 0; i < s.size(); ++i){
            if (count.find(s[i]) == count.end()) count.insert(pair<char, int>(s[i], 1));
            else count[s[i]] += 1;
        }
        for (size_t i = 0; i < s.size(); ++i){
            if (count[s[i]] == 1) return i;
        }
        return -1;
    }
};

int main(){
    string s = "loveleetcode";
    Solution solu;
    cout << solu.firstUniqChar(s) << endl;
}

File: python/n387_FirstUniq.py
# https://leetcode-cn.com/problems/first-unique-character-in-a-string/
class Solution:
    def firstUniqChar(self, s: str) -> int:
        index_list = []
        for i in range(len(s)):
            if i == s.index(s[i]):
                index_list.append(i)
            elif s.index(s[i]) in index_list:
                index_list.remove(s.index(s[i]))
        if not index_list:
            return -1
        else:
            return index_list[0]
            
    def firstUniqChar_a(self, s: str) -> int:
        index_map = {}
        for i, j in enumerate(s):
            if j not in index_map.keys():
                index_map[j] = 1
            else:
                index_map[j] += 1
        for k in s:
            if index_map[k] == 1:
                return s.index(k)
        return -1
            
            
if __name__ == "__main__":
    solu = Solution()
    s = "loveleetcode"
    print(solu.firstUniqChar_a(s))